code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/AbstractTokenUtil.java

```diff
@@ -396,4 +397,53 @@ public static boolean isUnrestrictedAccess(String accessMode){
         return UNRESTRICTED_ACCESSMODE.equalsIgnoreCase(accessMode);
     }
 
+    public Identity jsonToIdentity(Map<String, Object> jsonData) {
+        String externalId = ObjectUtils.toString(jsonData.get("externalId"));
+        String externalIdType = ObjectUtils.toString(jsonData.get("externalIdType"));
+        String name = ObjectUtils.toString(jsonData.get("name"));
+        String profilePicture = ObjectUtils.toString(jsonData.get("profilePicture"));
+        String profileUrl = ObjectUtils.toString(jsonData.get("profileUrl"));
+        String login = ObjectUtils.toString(jsonData.get("login"));
+        return new Identity(externalIdType, externalId, name, profileUrl, profilePicture, login);
+    }
+
+    public Token getUserIdentityFromJWT() {
+        Token token = new Token();
+        Map<String, Object> jsonData = getJsonData();
+        if (jsonData == null) {
+            return null;
+        }
+        String key = getJWT();
+        if (StringUtils.isNotBlank(key)) {
+            AuthToken authToken= authTokenDao.getTokenByKey(getJWT());
+            String tokenId = (String)ApiContext.getContext().getIdFormatter().formatId(objectManager.getType(Token.class), authToken.getId());
+            token.setId(tokenId);
+        }
+
+        Object idObject = jsonData.get(USER_IDENTITY);
+        if (idObject != null) {
+            Map<String, Object> idMap = CollectionUtils.toMap(idObject);
+            Identity userIdentity = jsonToIdentity(idMap);
+            String userType = ObjectUtils.toString(jsonData.get(USER_TYPE), null);
+            token.setUserIdentity(userIdentity);
+            token.setUserType(userType);
+        }
+        return token;
+    }
+
+    public Token retrieveCurrentToken() {
+        Token token = new Token();
+
+        if (findAndSetJWT()) {
+            Token userToken = getUserIdentityFromJWT();
+            if(userToken != null) {
+                token.setUserIdentity(userToken.getUserIdentity());
+                token.setUserType(userToken.getUserType());
+                token.setId(userToken.getId());
+                token.setJwt(getJWT());
+            }
+        }
+        log.info("retrieveCurrentToken returning {}", token);
```

Comment:
@cjellick Oh, yes that makes sense. I had just copied this method as is from ExternalServiceAuthProvider. I'll change it to debug level then